06 - Type Ahead/index-START.html

```diff
@@ -15,8 +15,68 @@
     </ul>
   </form>
 <script>
-const endpoint = 'https://gist.githubusercontent.com/Miserlou/c5cd8364bf9b2420bb29/raw/2bf258763cdddd704f8ffd3ea9a3e81d25e2c6f6/cities.json';
+  //// global variables
+  const endpoint = 'https://gist.githubusercontent.com/Miserlou/c5cd8364bf9b2420bb29/raw/2bf258763cdddd704f8ffd3ea9a3e81d25e2c6f6/cities.json';
+  let cities = [];
+  let searchTerm = '';
 
+  //// selectors
+  const form = document.querySelector('.search-form');
+  const input = document.querySelector('.search');
+  const suggestions = document.querySelector('.suggestions');
+
+  //// callbacks
+  // fetchData() is async if we decide to also fetch on change or on submit for example
+  const fetchData = async (URL) => {
+    try {
+      const res = await fetch(URL);
+      const data = await res.json();
+      cities = [...data];
+    } catch (error) {
+      const isError = () => {
+        const message = `<h2 class="errorMessage">${error}</h2>`;
+        suggestions.innerHTML = message;  
+      }
+      isError();
+    }
+  }
+
+  fetchData(endpoint);
+
+  const findMatch = (wordToSearch, dataToFilter) => {
+    return dataToFilter.filter(item => {
+      const regex = new RegExp(wordToSearch, 'gi');
+      return item.city.match(regex) || item.state.match(regex);
+    });
+  };
+
+  const displayResults = (evt) => {
+    evt.preventDefault();
+    searchTerm = evt.target.value;
+   
+    // if the input value is deleted reset the suggestions otherwise run findMatch()
+    const results = !searchTerm ? [] : findMatch(searchTerm, cities);
+    
+    const resultsToDisplay = results.map(result => {
+      const regex = new RegExp(searchTerm, 'gi');
+      const cityName = result.city.replace(regex, `<span class="hl">${searchTerm}</span>`);
```

Comment:
I am pretty sure there is a semantic html element for highlighting words.

Reply:
Tom you are a genius! completely forgot about it from my HTML intro classes. I replaced it with <mark> tags and styled it with the Wes color.